### ISSUE TITLE
BUG: Fetching service WSLD definition using metaservices add-on fails with certificate validation error when the server providing the service uses a SSL certificate that is not signed by a well-known certificate authority. 

### ISSUE THREAD

**Issue by petkivim** [vrk-kpa/xroad-joint-development#56]
**Affected components**:  metaservices add-on
**Affected documentation**: -
**Estimated delivery**: -
**External reference**: https://jira.csc.fi/browse/PVAYLADEV-89, https://jira.csc.fi/browse/PVAYLA-28

**Problem**

Fetching service WSLD definition using metaservices add-on fails with certificate validation error when the server providing the service uses a SSL certificate that is not signed by a well-known certificate authority.

Scenario:

• A service provider is using HTTPS protocol and SSL certificate verification is enabled.
• The service provider SSL certificate is added to the Internal SSL certificates (as required by the verification).
• The SSL certificate is self-signed or issued by some internal certification authority that is not present in the default Java trust store.

Using the service works as expected but trying to fetch the corresponding WSDL definition using the metaservice fails with the following error: 

```
<SOAP-ENV:Envelope>
<SOAP-ENV:Body>
<SOAPENV:Fault>
<faultcode>Server.ServerProxy.IOError</faultcode><faultstring>sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target</faultstring>
<detail>
<xroad:faultDetail>javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1937)
...
at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1512)
at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1440)
at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:338)
at ee.ria.xroad.proxy.serverproxy.MetadataServiceHandlerImpl.getWsdl(MetadataServiceHandlerImpl.java:221)
at ee.ria.xroad.proxy.serverproxy.MetadataServiceHandlerImpl.handleGetWsdl(MetadataServiceHandlerImpl.java:170)
at ee.ria.xroad.proxy.serverproxy.MetadataServiceHandlerImpl.startHandling(MetadataServiceHandlerImpl.java:95)
at ee.ria.xroad.proxy.serverproxy.ServerMessageProcessor.handleRequest(ServerMessageProcessor.java:163)
at ee.ria.xroad.proxy.serverproxy.ServerMessageProcessor.process(ServerMessageProcessor.java:90)
at ee.ria.xroad.proxy.serverproxy.ServerProxyHandler.handle(ServerProxyHandler.java:55)
...
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:145)
at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:131)
at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
...
</xroad:faultDetail>
</detail>
</SOAP-ENV:Fault>
</SOAP-ENV:Body>
</SOAP-ENV:Envelope> 
```

The problem can be fixed by installing the certificate in the Java certificate trust store but that requires shell (command line) access with root permissions to the security server host and restarting the security server. In addition, this behavior is inconsistent since calling the service via the proxy doest not require that the the certificate is present in the Java default trust store. It would be preferable that the metaservice would honor the same security settings as the proxy and would trust the same internal SSL certificates. 

**Acceptance criteria**
-  metaservices honor the same security settings as the proxy and trust the same internal SSL certificates

**Comment by olli-lindgren** [comment]
Current solution (in review) extracts the code creating the Apache HttpClient to a separate class, and that class is re-used in MetadataServiceHandlerImpl, replacing the HttpUrlConnection.

Affected components:
- proxy/ServerProxy (Refactored, logic unaffected)
- addons/metaservice/MetadataServiceHandler/getWsdl (security requirements now same as ServerProxy)

**Comment by hanhaka** [comment]
Merged code changes to develop branch in VRK-KPA repository via commit:
https://github.com/vrk-kpa/xroad/pull/33

**Comment by hanhaka** [comment]
Commited to XM/develop (finnish-6.9.0 pull request). Will be available with 6.9 release.
